prometheus.lua

```diff
@@ -90,10 +150,12 @@ local function full_metric_name(name, label_names, label_values)
   end
   local label_parts = {}
   for idx, key in ipairs(label_names) do
-    local label_value = (string.format("%s", label_values[idx])
-      :gsub("[^\032-\126]", "")  -- strip non-printable characters
-      :gsub("\\", "\\\\")
-      :gsub('"', '\\"'))
+    local label_value = ""
+    if type(label_values[idx]) == "string" and validate_utf8_string(label_values[idx]) then
```

Comment:
Currently all users of this library are forced to only use printable characters in their label values, so it's clearly the most common case.

As a performance optimization I would like to suggest checking whether a string consists of 7-bit characters, and only call `validate_utf8_string` if it does not. Checking a string against `find("[\128-\255]")` is about 5 times faster than running it through your implementation of validate_utf8_string.

Something like this:
```lua
local label_value = label_values[idx]
if label_value:find("[\128-\255]") then
  valid, idx = validate_utf8_string(label_value)
  if not valid then
    label_value = string.sub(label_value, 1, idx-1)
  end
end
label_value = label_value:gsub("\\", "\\\\"):gsub('"', '\\"')
```

Reply:
Hello,  I just tested strings contains only ascii characters on my local laptop and result is `validate_utf8_string` is still faster than `string.find`
Could show me your benchmark code?

Reply:
Ah, thanks for double checking this. [Here's the updated benchmark](https://github.com/knyar/utf8_validator.lua/blob/master/benchmark.lua), but note that I initially ran it using a standard lua interpreter, not OpenResty.

I seems that resty-specific lua optimizations do make your code faster than `string.find`:

```bash
$ lua benchmark.lua
benchmark for: very short ascii string, string 'a'
- validate took 383.985996, result true
- validate2 took 289.080858, result true
- ascii_only took 207.489967, result true
benchmark for: somewhat longer ascii string, string 'this is a much longer ascii string, hello'
- validate took 7672.196865, result true
- validate2 took 6624.720097, result true
- ascii_only took 753.315210, result true
benchmark for: utf8 stuff, string '¢€𤭢'
- validate took 3016.548157, result true
- validate2 took 2552.010059, result true
- ascii_only took 201.677799, result false
benchmark for: mixed stuff, string 'Pay in €. Thanks.'
- validate took 3876.733065, result true
- validate2 took 3433.573961, result true
- ascii_only took 305.641890, result false

$ resty benchmark.lua
benchmark for: very short ascii string, string 'a'
- validate took 65.901995, result true
- validate2 took 0.577927, result true
- ascii_only took 50.730944, result true
benchmark for: somewhat longer ascii string, string 'this is a much longer ascii string, hello'
- validate took 2144.397974, result true
- validate2 took 74.812174, result true
- ascii_only took 514.914989, result true
benchmark for: utf8 stuff, string '¢€𤭢'
- validate took 1522.389889, result true
- validate2 took 54.584980, result true
- ascii_only took 47.739983, result false
benchmark for: mixed stuff, string 'Pay in €. Thanks.'
- validate took 1377.006769, result true
- validate2 took 51.558018, result true
- ascii_only took 133.812189, result false
```

Given that OpenResty is the recommended way to run nginx+lua, I think we should optimize for that.

Thanks, please ignore my suggestion to use `string.find` as a preliminary check.

Reply:
Yep, luajit is the common scene